MaxText/maxtext_utils.py

```diff
@@ -0,0 +1,417 @@
+    with open(metrics_filename, 'w', encoding="utf8") as metrics_for_gcs:
+      for metrics_step in running_metrics:
+        metrics_for_gcs.write(str(json.dumps(metrics_step))+'\n')
+
+    metrics_for_gcs.close()
+    gcs_filename=os.path.join(config.metrics_dir, metrics_filename)
+    command = ["gsutil", "mv", metrics_filename, gcs_filename]
+    max_logging.log(f"Moving file {metrics_filename} to GCS...")
+    subprocess.run(command, check=True, capture_output=True)
+    max_logging.log(f"File {metrics_filename} moved successfully!")
+    running_metrics = [] # reset running_metrics to empty list
+  return running_metrics
+
+def _prepare_metrics_for_json(metrics, step, run_name):
+  """Converts metric dictionary into json supported types (e.g. float)""" 
+  metrics_dict = {}
+  for val in metrics['scalar']:
+    metrics_dict[val] = float(metrics['scalar'][val])
+  metrics_dict['step'] = float(step)
+  metrics_dict['run_name'] = run_name
+  return metrics_dict
+
+def setup_initial_state(model, tx, config, rng, mesh, checkpoint_manager):
+  """ We initialize the model and optimizer state, and optionally load from a
+  checkpoint as necessary.
+
+  Args:
+    model: the flax model to initialize
+    tx: the optax.GradientTransformation
+    config: config object
+    rng: jax.prng key
+    mesh: jax.devices() mesh
+    checkpoint_manager: an Orbax checkpointing.CheckpointManager object
+
+  Returns:
+    state: the initialized train state
+    state_mesh_annotations: the mesh annotations for the train state
+  """
+
+  unboxed_abstract_state, state_mesh_annotations = get_abstract_state(model, tx, config, rng, mesh)
+
+  # Initialization
+  with nn_partitioning.axis_rules(config.logical_axis_rules):
+    state, raw_params = checkpointing.load_state_if_possible(checkpoint_manager,
+                                                config.load_parameters_path,
+                                                config.load_from_other_directory,
+                                                config.load_from_other_directory_step,
+                                                unboxed_abstract_state,
+                                                mesh,
+                                                state_mesh_annotations)
+
+    state_mesh_shardings = jax.tree_map(
+        lambda p: jax.sharding.NamedSharding(mesh, p), state_mesh_annotations)
+    if not state:
+      init_train_state_partial = functools.partial(init_train_state, model, tx, config)
+      state = jax.jit(
+          init_train_state_partial,
+          in_shardings=None,
+          out_shardings=state_mesh_shardings
+      )(rng)
+      if raw_params: # If we loaded a partial state, we need to merge it.
+        state = state.replace(params = raw_params)
+    raw_params = None
+
+  state = unbox_logicallypartioned_trainstate(state)
+  return state, state_mesh_annotations
+
+def create_device_mesh(config, devices=jax.devices(), logging=True):
```

Comment:
Its possible this should be moved into shared I am not sure how generalizable this is - currently data/ fsdp/tp seem general but we may add pipeline parallelism, and maybe other models want different meshes, I am uncertain, so decided to move here until the point where some other models wants an identical sharding and can refactor this to max_utils